org.eclipse.xtext.common.types.ui/src/org/eclipse/xtext/common/types/ui/notification/JavaBuilderState.java

```diff
@@ -0,0 +1,259 @@
+					if (packageName.equals(typePackageName)) {
+						String simpleTypeName = qualifiedPath.lastSegment().toString();
+						qualifiedTypeNames.addAll(
+								getQualifiedTypeNames((String) typeLocator, packageName, simpleTypeName, packageFragment.getJavaProject()));
+					}
+				}
+			}
+		}
+
+		return qualifiedTypeNames;
+	}
+
+	/**
+	 * @return The {@code TypeNames} which are direct children of a given CompilationUnit.
+	 */
+	protected TypeNames _getQualifiedTypeNames(ICompilationUnit compilationUnit) {
+		return getQualifiedTypeNames(getTypeLocator(compilationUnit), getPackageName(compilationUnit),
+				getSimplePrimaryTypeName(compilationUnit), compilationUnit.getJavaProject());
+	}
+
+	private IPackageFragmentRoot getPackageFragmentRoot(IJavaElement javaElement) {
+		IJavaElement parent = javaElement.getParent();
+		return parent instanceof IPackageFragmentRoot ? ((IPackageFragmentRoot) parent) : getPackageFragmentRoot(parent);
+	}
+
+	private TypeNames getQualifiedTypeNames(String typeLocator, String packageName, String simpleName, IJavaProject project) {
+		TypeNames qualifiedTypeNames = new TypeNames(project);
+		String primaryTypeFqn = getQualifedTypeName(packageName, simpleName);
+
+		char[][] typeNames = null;
+		if (state != null) {
+			typeNames = state.getDefinedTypeNamesFor(typeLocator);
+		}
+
+		if (typeNames == null) {
+			TypeNames newTypeNames = new TypeNames(project);
+			newTypeNames.addTypeName(primaryTypeFqn, primaryTypeFqn);
+			return newTypeNames;
+		}
+
+		for (char[] typeName : typeNames) {
+			String qualifiedTypeName = getQualifedTypeName(packageName, new String(typeName));
+			qualifiedTypeNames.addTypeName(qualifiedTypeName, primaryTypeFqn);
+		}
+
+		return qualifiedTypeNames;
+	}
+
+	private String getQualifedTypeName(String packageName, String simpleTypeName) {
+		return packageName == null ? simpleTypeName : packageName + "." + simpleTypeName;
+	}
+
+	private String getTypeLocator(ICompilationUnit compilationUnit) {
+		IResource resource = compilationUnit.getResource();
+		if (resource != null && resource.getProjectRelativePath() != null) {
+			return resource.getProjectRelativePath().toString();
+		}
+		return null;
+	}
+
+	private String getPackageName(ICompilationUnit compilationUnit) {
+		IJavaElement parent = compilationUnit.getParent();
+		if (parent instanceof IPackageFragment && !((IPackageFragment) parent).isDefaultPackage()) {
+			return ((IPackageFragment) parent).getElementName();
+		}
+		return null;
+	}
+
+	private String getSimplePrimaryTypeName(ICompilationUnit compilationUnit) {
+		String elementName = compilationUnit.getElementName();
+		return elementName.substring(0, elementName.lastIndexOf("."));
+	}
+
+	private SimpleLookupTable getReferences() {
+		if (references != null) {
+			return references;
+		}
+
+		if (state == null) {
+			return null;
+		}
+
+		Object referencesField = readField(state, "references", null);
```

Comment:
This local variable is probably not necessary. `return (SimpleLookupTable) readField(..)` should be fine.